src/7.1_HW/Department.ts

```diff
@@ -0,0 +1,77 @@
+import {Employee, Status} from "./Employee";
+import {PreHiredEmployee} from "./PreHiredEmployee";
+
+export enum DepartmentDomainEnum {
+    ACCOUNTING = 'Accounting',
+    HR = 'HR',
+    IT = 'IT',
+}
+
+const ZERO_BUDGET_VALUE = 0;
+
+export class Department {
+    name: string;
+    domain: DepartmentDomainEnum;
+    employees: Employee[];
+    budget: Budget = {
+        debit: ZERO_BUDGET_VALUE,
+        credit: ZERO_BUDGET_VALUE
+    };
+
+    constructor(name: string, domain: DepartmentDomainEnum, employees: Employee[], budget: Budget) {
+        this.name = name;
+        this.domain = domain;
+        this.employees = employees;
+        this.budget = budget;
+    }
+
+    calculateBalance(): number {
+        return this.budget.debit - this.budget.credit;
+    }
+
+    addEmployee(newcomer: Employee | PreHiredEmployee): void {
+        if (newcomer instanceof Employee) {
+            this.employees.push(newcomer);
+            this.budget.debit -= newcomer.salary;
+        } else {
+            const employee = new Employee(newcomer.firstName, newcomer.lastName, newcomer.bankAccountNumber, newcomer.salary, Status.ACTIVE, this);
+            this.employees.push(employee);
+            this.budget.debit -= employee.salary;
+        }
+
+    }
+
+    hirePreHiredOrRegularEmployee(employeeOrPreHired: Employee | PreHiredEmployee): Employee {
+        if (this.isPreHiredEmployee(employeeOrPreHired)) {
+            const employee = new Employee(
+                employeeOrPreHired.firstName,
+                employeeOrPreHired.lastName,
+                employeeOrPreHired.bankAccountNumber,
+                employeeOrPreHired.salary,
+                Status.ACTIVE,
+                this
+            );
+            this.addEmployee(employee);
+            return employee;
+        } else {
+            this.addEmployee(employeeOrPreHired);
+            return employeeOrPreHired;
+        }
+    }
+
+    private isPreHiredEmployee(employee: Employee | PreHiredEmployee): employee is PreHiredEmployee {
```

Comment:
It's better to implement type guards outside of any instance. It will allow you to use them everywhere you might possibly need.